### ISSUE TITLE
doesn't seem to work on my computer for some weird reason

### ISSUE THREAD

**Issue by max-mapper** [sindresorhus/shell-path#4]
running inside an electron app (0.34.1) packaged with electron-packager, running on Yosemite

<img width="1035" alt="screen shot 2015-10-27 at 10 20 00 am" src="https://cloud.githubusercontent.com/assets/39759/10766352/699c8b30-7c94-11e5-9160-a5cf64f1ddb7.png">

note the last two commands didn't seem to inherit my $PATH. I'll try and figure out why it isn't working...

**EDIT** also tried setting `gid`, `uid`, and `cwd` to my user, group and home folder but that didn't change the output

**2nd EDIT** exec `whoami` returns `max`, which is my user. exec `sudo -Hiu max env` returns the correct env variables, including my correct PATH. so I can access it that way. just not through `echo $PATH`

pretty sure this used to work in older electrons. maybe something changed

**Comment by silverwind** [comment]
Does adding `'-i'` to the arguments in the last lines do anything?

Looks like we've forgot to include `-i` in `sp.sync` :worried:

**Comment by max-mapper** [comment]
@silverwind nope, but `sudo -Hiu max env` can access my PATH:

<img width="1035" alt="screen shot 2015-10-27 at 10 41 29 am" src="https://cloud.githubusercontent.com/assets/39759/10766999/575544fa-7c97-11e5-8922-272c545dd2f0.png">

**Comment by silverwind** [comment]
> `sudo -Hiu max env`

Interesting. Is that the minimum amount of arguments needed for it to work? Not sure we can or should use `sudo` just to get `PATH`.

**Comment by max-mapper** [comment]
Im not sure familiar with that command, but it doesn't actually require sudo permissions to run

**Comment by max-mapper** [comment]
from `man sudo`:

```

       -H          The -H (HOME) option sets the HOME environment variable to the homedir of the target user (root by default) as specified in passwd(5).  The
                   default handling of the HOME environment variable depends on sudoers(5) settings.  By default, sudo will set HOME if env_reset or
                   always_set_home are set, or if set_home is set and the -s option is specified on the command line.

       -i [command]
                   The -i (simulate initial login) option runs the shell specified in the passwd(5) entry of the target user as a login shell.  This means that
                   login-specific resource files such as .profile or .login will be read by the shell.  If a command is specified, it is passed to the shell for
                   execution.  Otherwise, an interactive shell is executed.  sudo attempts to change to that user's home directory before running the shell.  It
                   also initializes the environment, leaving DISPLAY and TERM unchanged, setting HOME, MAIL, SHELL, USER, LOGNAME, and PATH, as well as the
                   contents of /etc/environment on Linux and AIX systems.  All other environment variables are removed.

       -u user     The -u (user) option causes sudo to run the specified command as a user other than root.  To specify a uid instead of a user name, use #uid.
                   When running commands as a uid, many shells require that the '#' be escaped with a backslash ('\').  Note that if the targetpw Defaults option
                   is set (see sudoers(5)) it is not possible to run commands with a uid not listed in the password database.

```

**Comment by silverwind** [comment]
If it doesn't require actual elevation, that's fine. I didn't even know `sudo` had such functionality. Some of that might be OS X specific. Thanks for finding this!

**Comment by max-mapper** [comment]
oh yea `echo $PATH` works too. you'd just need to run `whoami` first to get the username to use (or use some other method)

<img width="542" alt="screen shot 2015-10-27 at 10 50 13 am" src="https://cloud.githubusercontent.com/assets/39759/10767278/9b1ab6b0-7c98-11e5-95d5-d3fb122fe9cb.png">

**Comment by silverwind** [comment]
Awesome! This looks to works well on OS X:

``` js
child_process.execFileSync('sudo', ['-Hiu', process.env.USER, 'echo $PATH']).toString()
```

Looks like `sudo` on Linux supports these flags too, thought my shell seems to have some issues, probably needs some escaping (I use zsh on both Linux and OS X):

``` js
> child_process.execFileSync('sudo', ['-Hiu', process.env.USER, 'echo $PATH']).toString()
zsh:1: no such file or directory: echo /usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
Error: Command failed: sudo -Hiu silverwind echo $PATH
```

**Comment by silverwind** [comment]
Might as well parse it out of `env` so it works on both Linux and OS X:

``` js
/PATH=(.+)\n/.exec(require('child_process').execFileSync('sudo', ['-Hiu', process.env.USER, 'env']).toString() || [])[1]
```

Edit: I see slightly different results using both methods on Ubuntu:

``` js
> /PATH=(.+)\n/.exec(require('child_process').execFileSync('sudo', ['-Hiu', process.env.USER, 'env']).toString() || [])[1].trim()
'/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin'
```

``` js
> child_process.execFileSync('/bin/sh', ['-i', '-c', 'echo $PATH']).toString().trim()
'/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/bin/X11:/usr/games'
```

``` js
> process.env.PATH
'/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/bin/X11:/usr/games'
```

I'll play in Electron to find out which method works best.

**Comment by davej** [comment]
Cross-referencing with https://github.com/sindresorhus/fix-path/issues/4
Looks to be the same issue.

**Comment by maaarghk** [comment]
I'm seeing this as well. I'm using zsh but it's in process.env.SHELL so I don't think that should be the issue. Attached is some debugging. 

![screen shot 2015-10-29 at 15 35 50](https://cloud.githubusercontent.com/assets/3168648/10823291/dfb90f90-7e52-11e5-95d8-61618830b28a.png)

**Comment by silverwind** [comment]
@maaarghk your issue with the sync method is fixed by https://github.com/sindresorhus/shell-path/commit/2e8c3f3c24fb5443f2b5944de7aa75f5ca8cb533, I'll push a release with that.

**Comment by maaarghk** [comment]
I was literally just about to come here and say that I noticed that missing flag :) Thanks!

**Comment by silverwind** [comment]
Looks @sindresorhus has to publish 1.0.3, Seems like I'm missing npm access on this repo.

**Comment by sindresorhus** [comment]
@silverwind Published, and you do now.

**Comment by silverwind** [comment]
:+1:

**Comment by silverwind** [comment]
Updated master to use both methods of obtaining the path and returning the longest path found out of all methods. Had a weird issue with `execFileSync` on Linux where

``` js
child_process.execFileSync('sudo', ['-Hiu', process.env.USER, 'echo $PATH'], {encoding: 'utf8'})
```

did fail with a `zsh:1: no such file or directory`. Switching to

``` js
child_process.execSync('sudo -Hiu' + process.env.USER + ' echo $PATH', {encoding: 'utf8'})
```

remedied that. May be a node bug.

**Comment by silverwind** [comment]
Released 1.0.4 with this.
